src/System.ServiceModel.NetNamedPipe/src/System/ServiceModel/Channels/PipeConnection.cs

```diff
@@ -0,0 +1,576 @@
+                            }
+                        }
+
+                        if (!_isShutdownWritten)
+                        {
+                            shouldWriteEOF = true;
+                            _isShutdownWritten = true;
+                        }
+                    }
+                }
+
+                ValueTask writeValueTask = default;
+                ValueTask<int> readValueTask = default;
+                if (shouldWriteEOF)
+                {
+                    writeValueTask = StartWriteZeroAsync(cancellationToken);
+                }
+
+                if (shouldReadEOF)
+                {
+                    readValueTask = StartReadZeroAsync(cancellationToken);
+                }
+
+                // wait for shutdown write to complete
+                if (shouldWriteEOF)
+                {
+                    try
+                    {
+                        await WaitForWriteZero(writeValueTask, timeout, true);
+                    }
+                    catch (TimeoutException e)
+                    {
+                        throw DiagnosticUtility.ExceptionUtility.ThrowHelper(
+                            new TimeoutException(SR.PipeShutdownWriteError, e), ExceptionEventType);
+                    }
+                }
+
+                // ensure we have received EOF signal
+                if (shouldReadEOF)
+                {
+                    try
+                    {
+                        await WaitForReadZero(readValueTask, timeout, true);
+                    }
+                    catch (TimeoutException e)
+                    {
+                        throw DiagnosticUtility.ExceptionUtility.ThrowHelper(
+                            new TimeoutException(SR.PipeShutdownReadError, e), ExceptionEventType);
+                    }
+                }
+                else if (existingReadIsPending)
+                {
+                    if (!TimeoutHelper.Wait(_atEOFEvent, timeoutHelper.RemainingTime()))
+                    {
+                        throw DiagnosticUtility.ExceptionUtility.ThrowHelper(
+                            new TimeoutException(SR.PipeShutdownReadError), ExceptionEventType);
+                    }
+                }
+                // else we had already seen EOF.
+
+                // at this point, we may get exceptions if the other side closes the handle first
+                try
+                {
+                    // write an ack for eof
+                    writeValueTask = StartWriteZeroAsync(cancellationToken);
+
+                    // read an ack for eof
+                    readValueTask = StartReadZeroAsync(cancellationToken);
+
+                    // wait for write to complete/fail
+                    await WaitForWriteZero(writeValueTask, timeout, false);
+
+                    // wait for read to complete/fail
+                    await WaitForReadZero(readValueTask, timeout, false);
```

Comment:
timeout doesn't look like it is used by these methods

Reply:
NVM, it is used in message.

Reply:
The caller won't know what's the appropriate exception message should be as we have different exception messages for socket etc. I could catch the OperationCanceledException, catch it and throw TimeoutException with the literal `{0}` where the timeout should be, have the caller catch TimeoutException, reformat the string with the timeout and then rethrow again. It seems wrong to throw 3 times to achieve this though. Do you have any thoughts about a more elegant way to solve this?